source/extensions/quic_listeners/quiche/envoy_quic_server_session.h

```diff
@@ -0,0 +1,164 @@
+// QUICHE uses offsetof().
+#pragma GCC diagnostic ignored "-Winvalid-offsetof"
+#pragma GCC diagnostic ignored "-Wtype-limits"
+
+#include "quiche/quic/core/http/quic_server_session_base.h"
+
+#pragma GCC diagnostic pop
+
+#include <memory>
+
+#include "common/common/logger.h"
+#include "common/common/empty_string.h"
+#include "common/network/filter_manager_impl.h"
+#include "common/stream_info/stream_info_impl.h"
+#include "envoy/network/connection.h"
+#include "envoy/event/dispatcher.h"
+#include "extensions/quic_listeners/quiche/envoy_quic_stream.h"
+#include "extensions/quic_listeners/quiche/envoy_quic_connection.h"
+
+namespace Envoy {
+namespace Quic {
+
+// Act as a Network::Connection to HCM and a FilterManager to FilterFactoryCb.
+class EnvoyQuicServerSession : public quic::QuicServerSessionBase,
+                               public Network::FilterManagerConnection,
+                               protected Logger::Loggable<Logger::Id::connection> {
+public:
+  // Owns connection.
+  EnvoyQuicServerSession(const quic::QuicConfig& config,
+                         const quic::ParsedQuicVersionVector& supported_versions,
+                         quic::QuicConnection* connection, quic::QuicSession::Visitor* visitor,
+                         quic::QuicCryptoServerStream::Helper* helper,
+                         const quic::QuicCryptoServerConfig* crypto_config,
+                         quic::QuicCompressedCertsCache* compressed_certs_cache,
+                         Event::Dispatcher& dispatcher);
+
+  // Overridden to delete connection object.
+  ~EnvoyQuicServerSession() override;
+
+  // Network::FilterManager
+  // Overridden to delegate calls to filter_manager_.
+  void addWriteFilter(Network::WriteFilterSharedPtr filter) override;
+  void addFilter(Network::FilterSharedPtr filter) override;
+  void addReadFilter(Network::ReadFilterSharedPtr filter) override;
+  bool initializeReadFilters() override;
+
+  // Network::Connection
+  void addConnectionCallbacks(Network::ConnectionCallbacks& cb) override;
+  void addBytesSentCallback(Network::Connection::BytesSentCb /*cb*/) override;
+  void enableHalfClose(bool enabled) override;
+  void close(Network::ConnectionCloseType type) override;
+  Event::Dispatcher& dispatcher() override { return dispatcher_; }
+  uint64_t id() const override {
+    // QUIC connection id can be 18 types. It's easier to use hash value instead
+    // of trying to map it into a 64-bit space.
+    return connection_id().Hash();
+  }
+  std::string nextProtocol() const override { return EMPTY_STRING; }
+  void noDelay(bool /*enable*/) override {
+    // No-op. TCP_NODELAY doesn't apply to UDP.
+  }
+  void setDelayedCloseTimeout(std::chrono::milliseconds timeout) override;
+  std::chrono::milliseconds delayedCloseTimeout() const override;
+  void readDisable(bool /*disable*/) override {
+    // Quic connection should be able to read through out its life time.
+    NOT_REACHED_GCOVR_EXCL_LINE;
```

Comment:
Again no need to do anything here but when we TODO-audit we could just have a mode where we drop rather than dispatch incoming packets for this session, just to avoid runtime crashes if someone utilizes this function in their own filters